x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -212,6 +220,24 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
             this.apiKeyAuthCache = CacheBuilder.<String, ListenableFuture<CachedApiKeyHashResult>>builder()
                 .setExpireAfterAccess(ttl)
                 .setMaximumWeight(maximumWeight)
+                .removalListener(notification -> {
+                    if (RemovalReason.EVICTED == notification.getRemovalReason() && getApiKeyAuthCache().count() >= maximumWeight) {
+                        evictionCounter.increment();
+                        logger.trace("API key with ID [{}] was evicted from the authentication cache, " +
+                                "possibly due to cache size limit", notification.getKey());
+                        final long previousCheck = lastEvictionCheckedAt.get();
+                        final long now = clock.instant().toEpochMilli();
+                        if (now - previousCheck >= EVICTION_CHECKING_INTERVAL && lastEvictionCheckedAt.compareAndSet(previousCheck, now)) {
+                            final long sum = evictionCounter.sum();
+                            evictionCounter.add(-sum); // reset by decrease
+                            if (sum >= EVICTION_WARNING_THRESHOLD) {
+                                logger.warn("Possible thrashing for API key authentication cache, " +
+                                    "[{}] eviction due to cache size within last [{}] seconds",
+                                    sum, EVICTION_CHECKING_INTERVAL / 1000);
+                            }
+                        }
+                    }
```

Comment:
Nit: I think this block is long enough to justify being a method rather than an in place lambda.

Reply:
I initially didn't extract it because this code needs the local variable `maximumWeight`. This means the extracted method has to return a lambda. Anyway, I went ahead and extracted it. The preamble of this class is already long enough and we can save some lines here.